### ISSUE TITLE
Self-host bootstrap does not initialize cljs.user ns

### ISSUE THREAD

**Issue by superstructor** [thheller/shadow-cljs#744]
Firstly, just want to say having worked with the boostrap target and compiler environment provided by shadow-cljs it is awesome work! Its a massive step up from the previous status quo in terms of the performance, analysis cache file generation and so forth!

## Background

[`shadow.cljs.bootstrap.browser/init`](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L169) fn takes a callback fn argument `init-cb`.

## Expected Behavior

I expect the `init-cb` callback fn to be called after the bootstrapped compiler is initialized.

## Actual Behavior

**Sometimes** the `init-cb` callback fn is called before the bootstrapped environment is completely ready. For example, this presents itself in our tests as the `js/window.cljs.user` JS object backing the default `'cljs.user` ns not being created in time for evaluating a `def` like:
```
(def x 1)
```
Resulting in a JavaScript exception during the evaluation phase:
```
Cannot set property 'x' of undefined
```
Because `js/window.cljs.core` is defined, but `js/window.cljs.user` is undefined.

On a browser test runner with shadow-cljs `:target :browser-test` this is an intermittent failure maybe 25% of fresh page loads. On a karma test runner with shadow-cljs `:target :karma` this fails 100% of the time. Introducing a delay after the `init-cb` fn is called resolves the issue 100% of the time. Therefore we have concluded it is a timing issue.

## Suspects

~1. That [`script/eval-js` in `execute-load!` fn](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L53) might be async so doesn't guarantee completion prior to `init-cb` callback fn being called.~
~2. Use of [`goog.async.run`](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L59) to implement the queue that calls the `init-cb` callback fn.~
3. Its not created. [shadow-cljs devtools gives the illusion that it is under development builds](https://github.com/thheller/shadow-cljs/issues/744#issuecomment-650142301).

**Comment by mike-thompson-day8** [comment]
@thheller 
As you have requested, we're putting together a repo to show the race condition but it might yet prove to be tricky to deliver a minimal example.  Maybe.  Anyway, we're trying and we'll be back soon on this point. 

In the meantime, I believe the issue is fairly clear to see: 
1.  The loading of bootstrap files happens via [BulkLoader](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L131). We assume this part is fine. 
2. And then each loaded file is added to the page via [a queued task](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L131).  And a [final queued task](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L131) is added at the very end which does the callback. 

For point 2 above to work correctly (ie. the callback happens after all scripts are loaded into the page), the assumption is that each queued task will be handled synchronously. But we don't think they are.  Hence a race condition, because the callback task is actioned before the work in the tasks ahead of it is completed. 

Afterall, the implementation of `queue-task!` is [clearly async](https://github.com/thheller/shadow-cljs/blob/3f98c42c69e239952f5c68824664c30c0b638f47/src/main/shadow/cljs/bootstrap/browser.cljs#L55-L59)

And even if `queue-task!` wasn't async, the loading of the javascript into the page is going to be async, even if it is inlined within an added <script> element.

**Comment by thheller** [comment]
The loading is done async in a run-queue. It is processed in order. Each task itself is not async so they will complete before the next task runs.

I'm not ruling out that there is an issue somewhere. Nobody else has reported this before and all this is used for years now by `maria.cloud`. I still would like a see an example of your setup. You mention `:karma` or `:browser-test` so I assume you are using a custom runner namespace? You mention 100% failure rate in `:karma` so this shouldn't be that hard to reproduce?

**Comment by superstructor** [comment]
Thanks to some extensive pair debugging with @mike-thompson-day8 I think we've worked it out.

This turned out to be a bit of a red herring due to sharing global state on the `js/window` object between multiple environments.

First, we have observed
1. `cljs.user` is considered a 'default/fallback ns' baked into a lot of the ClojureScript code (e.g. analyzer, compiler etc); and
2. `cljs.user` exists in `cljs.analyzer/namespaces` in compiler state at the time `init-cb` is called; yet
3. `js/window.cljs.user` JavaScript object is *not* created for the user until `js/eval("goog.provide(\"cljs.user\")` or equivalent is evaluated.

So, I repeat, `js/window.cljs.user` is **not** created in the simple case. `maria.cloud` probably has this bug. They just don't evaluate code in `cljs.user` by default as they use `maria.user` which by definition must be created by them which I assume will have a `goog.provide` that will in turn call `goog.exportPath_`  to create the js objects for `maria.user`.

Now the story gets complex. Because shadow-cljs development builds that run either a browser test runner (`:target :browser`) or an example app to test such a library under development include `shadow.cljs.devtools.client.browser` (the HUD display etc) that includes its own JS loader that loads and evaluates `goog.provide(\"cljs.user\")` for its own purposes.

So, because a completely independent asynchronous process just happens to create `cljs.user`, it can give the **illusion** that `cljs.user` is created by the time `init-cb` is called but of course there is no guarantee that it will finish before `init-cb` is called. Hence the race condition and intermittent failures in the browser test runner. Further, of course as soon as you take away `shadow.cljs.devtools.client.browser` in release builds those won't work if you evaluate code in the default `cljs.user` ns.

 In short, the fix was to `js/eval("goog.provide(\"cljs.user\")"` immediately after `init-cb` is called but before evaluating any code. 

Before we finally lay this issue to rest we should probably ask the question - should `js/window.cljs.user` be created for an API user of `shadow.cljs.bootstrap.browser/init` or is that ClojureScript's responsibility, or is it OK to expect the API user to just eval a `goog.provide` ? If the answer is that it should be created for us then this is a bug in the existing implementations that no one has noticed because they don't use the default `cljs.user` ns, or they use it under circumstances where it will be created by other side effects.

**Comment by superstructor** [comment]
Re your request for a test setup @thheller  I created [`day8/shadow-cljs-bootstrap-init-cb`](https://github.com/day8/shadow-cljs-bootstrap-init-cb) but it **never** creates `cljs.user` so I am missing something regarding the `shadow-cljs.edn` config that causes the devtools to eval the `goog.provide("cljs.user")` for that example. Our main project will be released as open source soon but its a lot of irrelevant code that would just confuse the issue so tried to break out into a simpler example.

**Comment by thheller** [comment]
I'll take a look later the karma setup isn't working for me.

As a guess however since you say the issue issue is with `cljs.user`? That is REPL namespace and thus is loaded lazily when the REPL first connects. It does not exist in self-hosted builds at all unless you create an actual `cljs.user` namespace on disk and add it to the precompiled namespaces and actually require it? Or if you compile it from self-hosted of course.

The "race condition" part maybe suggest that you were relying on the `watch` for the host-build having that ns loaded which of course isn't true in `karma` since `watch` isn't running.

**Comment by superstructor** [comment]
Yeah the point now is that - should `cljs.user` be created by the self-hosted ClojureScript / shadow-cljs bootstrap APIs @thheller ?

Because it exists in `cljs.analyzer/namespaces` in compiler state prior to you creating an actual namespace ? It's also assumed as a 'safe default' throughout the ClojureScript source ? e.g. in [`eval*`](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/js.cljs#L793)

If you setup a self-hosted build correctly, run `bootstrap/init` then evaluate `(def x 0)` you get an error because the default ns `cljs.user` exists in the analyzer state but the backing JS object is undefined.

Isn't that a bug ?

**Comment by thheller** [comment]
I'm honestly not sure. I have never done anything that actually used self-hosted in any meaningful way.

`cljs.user` exists in the analyzer data because of this:
https://github.com/clojure/clojurescript/blob/92707a0b1ffb2a9fb6c019f877008eaeaedf97e5/src/main/clojure/cljs/env.cljc#L48

Why that is a thing I do not know. I can initialize it in the `bootstrap/init` or you can do it on your end by evaling something like this first.
```
(ns cljs.user
  (:require [something.maybe :refer (foo)]))
```

Might make sense to ensure its properly created I guess. I'll see if I can find what the actual self-hosted code does.

**Comment by thheller** [comment]
Alright I did a quick check using regular CLJS to see what it does and it exhibits the exact same issue.

```clojure
(ns demo.app
  (:require [cljs.js :as cjs]
            [cljs.pprint :refer [pprint]]))

(enable-console-print!)

(def state (cjs/empty-state))

(defn js-eval [{:keys [source]}]
  (js/eval source))

(cjs/eval-str state "(def x 1)" "foo" {:eval js-eval} prn)
(cjs/eval-str state "(prn [:x x])" "foo" {:eval js-eval} prn)
```
via
```bash
$ clj -m cljs.main -O simple -c demo.app
```
Resulting in
```clojure
{:error #error {:message "ERROR", :data {:tag :cljs/analysis-error}, :cause #object[TypeError TypeError: Cannot set property 'x' of undefined]}}
{:error #error {:message "ERROR", :data {:tag :cljs/analysis-error}, :cause #object[TypeError TypeError: Cannot read property 'x' of undefined]}}
```

It also works from the REPL since that sneakily creates the `cljs.user` ns via 
https://github.com/thheller/clojurescript/blob/3a0c07477ae781bf521bdc2b074ed7b783bb93f3/src/main/cljs/clojure/browser/repl.cljs#L166.

So one option is to just always do that when initializing or just leave that on the user the eval something like `(ns cljs.user)` first. I don't really use self-hosted so I have no opinion on this.

**Comment by mike-thompson-day8** [comment]
Many thanks @thheller.  I'm glad we got to the bottom of this - it was really baffling.
